packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,255 @@
+      },
+      async deposit(_payment) {
+        // XXX deposit an ERTP payment to the remote account
+        Fail`deposit not implemented yet`;
+      },
+      async getPurse(_brand) {
+        Fail`getPurse not implemented yet`;
+      },
+      async close() {
+        // XXX retrieve all assets first?
+        // XXX do we also revoke the port?
+        return E(this.state.connection).close();
+      },
+    },
+  );
+
+/** @param {import('@agoric/zone').Zone} zone */
+const prepareConnectionHandler = zone =>
+  zone.exoClass(
+    'ConnectionHandler',
+    undefined, // XXX interface guard
+    /**
+     * @param {(
+     *   connection: Connection,
+     *   localAddr: string,
+     *   remoteAddr: string,
+     * ) => void} [onOpen]
+     *   optional cb handler for connection creation
+     * @param {(connection: Connection, reason: unknown) => void} [onClose]
+     *   optional cb handler for connection closing
+     * @param {(
+     *   connection: Connection,
+     *   bytes: Uint8Array,
+     * ) => PromiseVow<string>} [onReceive]
+     *   optional cb handler for connection recieving packet
+     */
+    (onOpen, onClose, onReceive) => ({ onOpen, onClose, onReceive }),
+    {
+      async onOpen(connection, localAddr, remoteAddr) {
+        trace(`ICA Channel Opened for ${localAddr} at ${remoteAddr}`);
+        const { onOpen } = this.state;
+        if (onOpen && typeof onOpen === 'function') {
+          void onOpen(connection, localAddr, remoteAddr);
+        }
+      },
+      async onClose(connection, reason) {
+        trace(`ICA Channel closed. Reason: ${reason}`);
+        const { onClose } = this.state;
+        if (onClose && typeof onClose === 'function') {
+          void onClose(connection, reason);
+        }
+      },
+      async onReceive(connection, bytes) {
+        const { onReceive } = this.state;
+        if (onReceive && typeof onReceive === 'function') {
+          void onReceive(connection, bytes);
+        }
+        return '';
+      },
+    },
+  );
+
+export const OrchestrationI = M.interface('Orchestration', {
+  provideAccount: M.callWhen(M.string(), M.string())
+    .optional(M.remotable('Port'))
+    .returns(M.remotable('ChainAccount')),
+  getChain: M.callWhen(M.string()).returns(M.remotable('Chain')),
+});
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<typeof prepareChainAccount>} createChainAccount
+ * @param {ReturnType<typeof prepareConnectionHandler>} makeConnectionHandler
+ */
+const prepareOrchestration = (
+  zone,
+  createChainAccount,
+  makeConnectionHandler,
+) =>
+  zone.exoClassKit(
+    'Orchestration',
+    {
+      admin: M.interface('OrchestrationAdmin', {
+        foo: M.call().returns(M.string()),
+      }),
+      public: OrchestrationI,
+    },
+    /** @param {Partial<OrchestrationPowers>} [initialPowers] */
+    initialPowers => {
+      /** @type {PowerStore} */
+      const powers = zone.detached().mapStore('PowerStore');
```

Comment:
why treat this as a collection? the keys are statically known. seems like using them as individual properties is more straightforward.

Reply:
oh... maybe because state properties can't be added on upgrade?
please add a comment to that effect, if that's the reason.

Reply:
There is now a comment with more details and an issue ref